lsp-mode.el

```diff
@@ -1877,6 +1897,101 @@ WORKSPACE is the workspace that contains the progress token."
+
+(defvar-local lsp--modeline-code-actions-string nil
+  "Holds the current code action string on modeline.")
+
+(declare-function all-the-icons-octicon "ext:all-the-icons")
+
+(defun lsp--modeline-code-actions-icon ()
+  "Build the icon for modeline code actions."
+  (if (featurep 'all-the-icons)
+      (all-the-icons-octicon "light-bulb"
+                             :face 'lsp-modeline-code-actions-face
+                             :v-adjust -0.0575)
+    (propertize "💡" 'face 'lsp-modeline-code-actions-face)))
+
+(defun lsp--modeline-build-code-actions-string (actions)
+  "Build the string to be presented on modeline for code ACTIONS."
+  (-let* ((icon (lsp--modeline-code-actions-icon))
+          (first-action-string (propertize (->> actions
+                                                lsp-seq-first
+                                                (gethash "title")
+                                                (replace-regexp-in-string "[\n\t ]+" " "))
+                                           'face 'lsp-modeline-code-actions-face))
+          (single-action? (= (length actions) 1))
+          (string (if single-action?
+                      (format " %s %s " icon first-action-string)
+                    (format " %s %s %s " icon first-action-string
+                            (propertize (format "(%d more)" (seq-length actions))
+                                        'display `((height 0.9))
+                                        'face 'lsp-modeline-code-actions-face)))))
+    (propertize string
+                'help-echo (concat "Apply code actions (s-l a a)\nmouse-1: "
+                                   (if single-action?
+                                       first-action-string
+                                     "select from multiple code actions"))
+                'mouse-face 'mode-line-highlight
+                'local-map (make-mode-line-mouse-map
+                            'mouse-1 (lambda ()
+                                       (interactive)
+                                       (if single-action?
+                                           (lsp-execute-code-action (lsp-seq-first actions))
+                                         (call-interactively #'lsp-execute-code-action)))))))
+
+(defun lsp-modeline--update-code-actions (actions)
+  "Update modeline with new code ACTIONS."
+  (when lsp-modeline-code-actions-kind-regex
+    (setq actions (seq-filter (-lambda ((&hash "kind"))
+                                (or (not kind)
+                                    (s-match lsp-modeline-code-actions-kind-regex kind)))
+                              actions)))
+  (if (seq-empty-p actions)
+      (setq-local global-mode-string (remove '(t (:eval lsp--modeline-code-actions-string)) global-mode-string))
+    (progn
+      (setq lsp--modeline-code-actions-string (lsp--modeline-build-code-actions-string actions))
+      (add-to-list 'global-mode-string '(t (:eval lsp--modeline-code-actions-string))))))
```

Comment:
this looks problematic: we are adding the string to the "global" `global-modeline-variable` but then we remove it from the "local" `global-modeline-variable`.

Reply:
Hum, will `add-to-list` set to the global variable?
Should I use `setq-local` with `append` instead so?

Reply:
yes.

Reply:
@yyoncho I confirmed that `add-to-list` here, change the local variable of the buffer, not the global one, so  I think it is ok to keep it this way, what do you think?

Reply:
> @yyoncho I confirmed that add-to-list here, change the local variable of the buffer, not the global one

If setq-local runs first it creates buffer-local variable. Before that add-to-list works on the global variable.